### PR TITLE
Static file uploading API

### PR DESCRIPTION
**Pull request by OmarIthawi**
An API for uploading static files.

 [(8) [Tahoe] Implement new API endpoint for fetching files from URL](https://trello.com/c/7SxUIKIP/3126-8-tahoe-implement-new-api-endpoint-for-fetching-files-from-url)

### CODE REVIEW

**Comment by thraxil** [comment]
Sorry it took me a while to get around to checking this one out.

I'm trying to fully understand the constraints here. I'm assuming that the images uploaded are not considered "sensitive" (ie, if a user who isn't logged in/authorized gets hold of the URL of an image or derived favicon and can view it, that's not a problem).

When Solution A "fetches those icons and stores them" I assume that would probably mean storing them to a cloud storage bucket, S3 bucket, or just a web-accessible directory somewhere that we could predict the URL based on the file stored (an aside, but I think any chance to store uploaded files in a bucket rather than a local filesystem behind nginx is one we should take). Why does the list of `['newUrl1,newUrl2,newUrl3, ...]` have to go all the way back to the frontend? If it's the django backend downloading the icons and storing them somewhere, wouldn't it also be able to just directly set them in the config JSON? Also, we push the download/store functionality to a celery job, it should be quite parallelizable. 

> Cannot clean up images since we don't know which image belongs to which site

Could we just include the site in the filename/URL of the stored image?

Overall, I'd probably go with something closer to Solution A rather than the sort of proxy/CDN approach in B. Django serving up files directly usually turns out badly.

**Comment by grozdanowski** [comment]
@thraxil the reason why the frontend needs to write the changes into the site configuration JSON is to avoid accidental overrides or data leaks. The user may have done multiple changes to the site ocnfiguration in AMC, but not saved yet. That means that there is a new configuration containing new JSON values stored in Redux waiting to be committed to the site configuration. So if the user uploads the favicons and the backend stores the values into the configuration, then the user later hits "Save", those favicon values will be overridden.
Also, we may want to utilise this API for more than one function, not just favicon fetching.

Site organisation as part of the filename is a good idea.

File: openedx/core/djangoapps/appsembler/sites/api.py [hunk 1]
```diff
@@ -68,18 +81,97 @@ def post(self, request, format=None):
...
+                }, {
+                    "source": "https://openclipart.org/download/292749/abstract-icon2.png",
+                    "dest": "https://tahoe.appsembler.com/customer_files/a12bc334fd/abstract.png"
+                }, {
+                    "source": "https://openclipart.org/download/292749/abstract-icon3.png",
+                    "dest": "https://tahoe.appsembler.com/customer_files/c334d1334df/abstract.png"
+                }]
+            }
+
+        Response on Error:
```
Comment (thraxil) [inline openedx/core/djangoapps/appsembler/sites/api.py]: At the moment it's bailing and returning an error on the first one that fails and that's fine. But eventually I think we'll want to be able to give better info back to client-side and handle things a little more robustly, potentially giving them error messages for multiple URLs. If we structure the error response that way now we could avoid changing the client-side later when we have a better implementation.

**Comment by OmarIthawi** [comment]
[(8) [Tahoe] Implement new API endpoint for fetching files from URL](https://trello.com/c/7SxUIKIP/3126-8-tahoe-implement-new-api-endpoint-for-fetching-files-from-url)

File: openedx/core/djangoapps/appsembler/sites/api.py [hunk 2]
```diff
@@ -68,18 +81,97 @@ def post(self, request, format=None):
...
+        if not (isinstance(urls, list) and urls):
+            return Response({
+                'success': False,
+                'value': 'No files were provided.',
+            }, status=status.HTTP_400_BAD_REQUEST)
+
+        stored_urls = []
+        for source_url in urls:
+            try:
+                response = requests.get(source_url)
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply openedx/core/djangoapps/appsembler/sites/api.py]: Done.

File: openedx/core/djangoapps/appsembler/sites/api.py [hunk 3]
```diff
@@ -68,18 +81,97 @@ def post(self, request, format=None):
...
+                "url": "https://openclipart.org/download/292749/abstract-icon3.png"
+            }
+    """
+
+    parser_classes = (JSONParser,)
+
+    def post(self, request):
+        storage = get_customer_files_storage()
+
+        urls = request.data.get('urls')
```
[the remark this answers is not in the document]
Comment (OmarIthawi) [reply openedx/core/djangoapps/appsembler/sites/api.py]: done

File: openedx/core/djangoapps/appsembler/sites/api.py (hunk 1 above)
Comment (OmarIthawi) [reply openedx/core/djangoapps/appsembler/sites/api.py]: Good idea, I think this is more useful as we put things in a celery task.

**Comment by OmarIthawi** [comment]
@thraxil this is ready for another review. I've added the important things that we need to do, but still, I think we'd eventually want to refactor the whole feature (including resizing) in a celery task once we have a complete feature.

**Comment by OmarIthawi** [comment]
Thanks @thraxil!

@grozdanowski this is ready for another review, alternatively, we can go ahead and merge it and fix other issues if it appears.

**Comment by OmarIthawi** [comment]
@grozdanowski I'm merging this and deploying to staging soon.

**Pull request merged**